src/GraphQL.Federation/Instrumentation/FederatedTraceBuilder.cs

```diff
@@ -0,0 +1,238 @@
+            public Node ToProto() => _root;
+
+            /// <summary>
+            /// Adds error to the root node of the proto tree.
+            /// </summary>
+            /// <param name="error">error to add</param>
+            public void AddRootError(ExecutionError error)
+            {
+                var root = GetOrCreateNode(ResultPath.ROOT_PATH);
+                var protoError = new Error
+                {
+                    Message = error.Message
+                };
+
+                foreach (var location in error.Locations)
+                {
+                    var l = new Location
+                    {
+                        Column = (uint)location.Column,
+                        Line = (uint)location.Line
+                    };
+                    protoError.Location.Add(l);
+                }
+                root.Error.Add(protoError);
+            }
+
+            /// <summary>
+            /// Adds field trace record to proto tree. 
+            /// </summary>
+            /// <param name="record">Record to add</param>
+            public void AddField(PerfRecord record)
+            {
+                var path = record.MetaField<IEnumerable<object>>("path").ToList();
+
+                var p = ResultPath.FromList(path);
+                var node = GetOrCreateNode(p);
+                var type = record.MetaField<string>("type");
+                var parentType = record.MetaField<string>("parentType");
+                var responseName = record.MetaField<string>("responseName");
+                var errors = record.MetaField<ExecutionErrors>("errors").ToArray();
+       
+                node.StartTime = (ulong)record.Start * 1000000;
+                node.EndTime = (ulong)record.End * 1000000;
+                node.ResponseName = responseName;
+                node.ParentType = parentType;
+                node.Type = type;
+                if (errors == null)
+                    return;
+                foreach (var error in errors)
+                {
+                    var protoError = new Error
+                    {
+                        Message = error.Message
+                    };
+
+                    foreach (var location in error.Locations)
+                    {
+                        var l = new Location
+                        {
+                            Column = (uint)location.Column,
+                            Line = (uint)location.Line
+                        };
+
+                        protoError.Location.Add(l);
+                    }
+
+                    node.Error.Add(protoError);
+                }
+            }
+
+            /// <summary>
+            /// Returns a <see cref="Node"/> represented by the given path if it has been calculated before.
+            /// If  not then it will create the node and its parents and store them and return the node represented
+            ///  by the path.
```

Comment:
```suggestion
            /// If not then it will create the node and its parents and store them and return the node represented
            /// by the path.
```